### ISSUE TITLE
question on  temperature loss "alpha" for discrete action

### ISSUE THREAD

**Issue by dbsxdbsx** [yining043/SAC-discrete#2]
According to paper: SOFT ACTOR-CRITIC FOR DISCRETE ACTION SETTINGS,  the loss formula (11) for temperature , there is a`pi_t(s_t)` at the very outside of the formula, which is action_probs for discrete actions. But according to the code here:
```
 # alpha loss for temperature parameter
    pi_entropy = -tf.reduce_sum(action_probs * log_action_probs, axis=-1)
    alpha_backup = tf.stop_gradient(target_entropy - pi_entropy)
    alpha_loss   = -tf.reduce_mean(log_alpha * alpha_backup)
```

seems that `log_action_probs` only affects `log_action_probs`, but not `target_entropy`, why?

**Comment by yining043** [comment]
Hi bro, 

This part of the code is forked from the repo [soft-actor-critich](https://github.com/ac-93/soft-actor-critic) so actually, I do not know much about this implementation. I've tried to email the original author to get his help. Will let you know if I got any updates.

**Comment by dbsxdbsx** [comment]
@yining043 , thanks.

**Comment by ac-93** [comment]
Hi, 

I think the notation in that paper may be slightly different from how I would write it.

The original loss equation for temperature parameter is given as 

![CodeCogsEqn](https://user-images.githubusercontent.com/55698603/90982070-aa970f80-e55c-11ea-8a50-77825c01d613.gif)

from the soft actor critic algorithms and applications paper (eq 18 https://arxiv.org/pdf/1812.05905.pdf).

For the discrete action setting pi_t is a categorical distribution and the expected value can be calculated as 

![CodeCogsEqn-2](https://user-images.githubusercontent.com/55698603/90982221-a61f2680-e55d-11ea-9a5e-33485b04b977.gif)

both \alpha and the target entropy are independent of the action distribution so this can be rewritten as 

![CodeCogsEqn-9](https://user-images.githubusercontent.com/55698603/90982712-d2887200-e560-11ea-8c3f-6337b1d233c5.gif)

and because we are using a categorical action distribution we have

![CodeCogsEqn-4](https://user-images.githubusercontent.com/55698603/90982346-83414200-e55e-11ea-8cd8-86563cf1fa39.gif)

therefore the original equation can be rewritten as 

![CodeCogsEqn-10](https://user-images.githubusercontent.com/55698603/90982774-5fcbc680-e561-11ea-96a5-417845d26739.gif)

Which is hopefully what the code does

A more intuitive way of understanding this equation is that we are trying to find alpha to minimize the difference between the current entropy of the policy, which can be calculated as  

![CodeCogsEqn-8](https://user-images.githubusercontent.com/55698603/90982620-33637a80-e560-11ea-94c3-01b1c4625790.gif)

(see https://en.wikipedia.org/wiki/Entropy_(information_theory))

and the target entropy H.

Hopefully I've not made any mistakes and this helps!

Best,
Alex

**Comment by dbsxdbsx** [comment]
@ac-93, thanks for your answer.  For the 1st explanation, I hope I'v understanded. If  I am wrong, tell me plz:
Both the equations from paper and here are correct, they are  equal to each other. 
But why equations of alpha_loss between code and paper are little different , when comparing pairs of value and policy loss? 
Because,  there is no calculation for prob of a "specific" action under state s_t, that is  the coefficient can be eliminated before H_hat, but NOT for the coefficient of log_action_prob(s_t).   And similiar explanation should be held for  value and policy loss.

But still, my intuition tell me that the very left coefficient "pi_t(s_t)^T" of equation 11 from paper should be innerside as the sole coefficient of log_action_Prob(s_t), so that it would not affect H_hat----Ok, this intuition maybe wrong due to my poor linear algebra.

**Comment by ac-93** [comment]
Sorry I'm not sure if I fully understand your question.

The reason we can move H_hat outside of the summation is because it is a constant and does not depend on the current action.

For an example 3 action policy and given state s_t we could have

pi        = [0.1, 0.4, 0.5]
log_pi = [-2.30258509, -0.91629073, -0.69314718]
H_hat = 0.05

in this case 

sum( pi * (log_pi - H_hat) ) = sum( pi * logpi ) - H_hat = 0.9933... 

and this should be the same for all cases.

If it helps I think you could rewrite the code to be closer to the paper, where target_entropy is included in the line with reduce_sum, I'm pretty sure the results will be the same though.

Best, 
Alex

**Comment by dbsxdbsx** [comment]
@ac-93 , thanks for answer again. So that is it. Both the equation from code and paper are correct. Just because H_hat is a CONSTANT,  so it doesn't need to multiple by a sum(pi) which equals 1.   
Now I fully understand.

By the way, I just found a paper which refers to your paper, called `Discrete and Continuous Action Representation for Practical RL in Video Games`. It combines both  discrete and continuous action in the same form. Though there is no code, I think it is a good idea.  And personally, for the state of art  algo SAC, I do hope  both discrete and continuous action could be in the same form, then I don't need to take 2 different models for  discrete and continuous  respectively 😀.

Maybe you would like it.

And thanks @yining043  again for your help.

**Comment by ac-93** [comment]
Glad I could help!

Would just like to point out that I'm not an author of the Discrete SAC paper, that's all https://github.com/p-christ as far as I know.

That is an interesting paper you just shared though, and I agree that a hybrid discrete/continuous action space is an interesting area of research!

Cheers,
Alex

**Comment by yining043** [comment]
Thanks, @ac-93 for your help! 👍
